File: 2022/day5/rust/src/main.rs
use std::{fs::File, io::BufReader};

use aoc22d5::{part1, part2, SolutionError};

fn main() -> Result<(), SolutionError> {
    let (part, path) = aoc22d5::parse_args();
    let part = if let Some("1") = part.as_deref() {
        1
    } else {
        2
    };
    let path = path.unwrap_or_else(|| String::from("../input.txt"));

    let file = File::open(&path)
        .unwrap_or_else(|e| panic!("Couldn't open the file at path {path}\n\nERR:\n{e}"));
    let mut reader = aoc22d5::InputParser::new(
        BufReader::new(file),
        if part == 1 {
            &part1::crane
        } else {
            &part2::crane
        },
    )?;

    reader.apply_operations()?;
    let answer = reader.read_tops();
    println!("{answer}");

    Ok(())
}

File: 2022/day5/ocaml/bin/main.ml
open Aoc22d5
open Common

let () =
    let part, path = match get_opt Sys.argv 1, get_opt Sys.argv 2 with
    | Some "1", Some path -> 1, path
    | Some _, Some path -> 2, path
    | Some "1", None -> 1, "../input.txt"
    | _ -> 2, "../input.txt" in

    let crane = if part == 1 then Part1.crane else Part2.crane in
    let ic = open_in path in
    try
        (match Common.solve crane ic with
        | Ok s -> s
        | Error e -> sexp_of_error e |> Core.Sexp.to_string_hum)
        |> print_endline;
        close_in ic
    with e ->
        close_in_noerr ic;
        raise e

